ledger/complete/wal/checkpoint_v6_reader.go

```diff
@@ -0,0 +1,722 @@
+	}
+
+	if actualSum != expectedSum {
+		return nil, 0, fmt.Errorf("invalid checksum in checkpoint header, expected %v, actual %v",
+			expectedSum, actualSum)
+	}
+
+	err = ensureReachedEOF(reader)
+	if err != nil {
+		return nil, 0, fmt.Errorf("fail to read checkpoint header file: %w", err)
+	}
+
+	return subtrieChecksums, topTrieChecksum, nil
+}
+
+// allPartFileExist check if all the part files of the checkpoint file exist
+// it returns nil if all files exist
+// it returns os.ErrNotExist if some file is missing, use (os.IsNotExist to check)
+// it returns err if running into any exception
+func allPartFileExist(dir string, fileName string, totalSubtrieFiles int) error {
+	matched, err := findCheckpointPartFiles(dir, fileName)
+	if err != nil {
+		return fmt.Errorf("could not check all checkpoint part file exist: %w", err)
+	}
+
+	// header + subtrie files + top level file
+	if len(matched) != 1+totalSubtrieFiles+1 {
+		return fmt.Errorf("some checkpoint part file is missing. found part files %v. err :%w",
+			matched, os.ErrNotExist)
+	}
+
+	return nil
+}
+
+// findCheckpointPartFiles returns a slice of file full paths of the part files for the checkpoint file
+// with the given fileName under the given folder.
+// - it return the matching part files, note it might not contains all the part files.
+// - it return error if running any exception
+func findCheckpointPartFiles(dir string, fileName string) ([]string, error) {
+	headerPath := filePathCheckpointHeader(dir, fileName)
+	pattern := headerPath + "*"
+	matched, err := filepath.Glob(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("could not find checkpoint files: %w", err)
+	}
+
+	// build a lookup with matched
+	lookup := make(map[string]struct{})
+	for _, match := range matched {
+		lookup[match] = struct{}{}
+	}
+
+	parts := make([]string, 0)
+	// check header exists
+	_, ok := lookup[headerPath]
+	if ok {
+		parts = append(parts, headerPath)
+		delete(lookup, headerPath)
+	}
+
+	// check all subtrie parts
+	for i := 0; i < subtrieCount; i++ {
+		subtriePath, _, err := filePathSubTries(dir, fileName, i)
+		if err != nil {
+			return nil, err
+		}
+		_, ok := lookup[subtriePath]
+		if ok {
+			parts = append(parts, subtriePath)
+			delete(lookup, subtriePath)
+		}
+	}
+
+	// check top level trie part file
+	toplevelPath, _ := filePathTopTries(dir, fileName)
+
+	_, ok = lookup[toplevelPath]
+	if ok {
+		parts = append(parts, toplevelPath)
+		delete(lookup, toplevelPath)
+	}
+
+	return parts, nil
+}
+
+var errCheckpointFileExist = errors.New("checkpoint file exists already")
+
+// noneCheckpointFileExist check if none of the checkpoint header file or the part files exist
+// it returns nil if none exists
+// it returns errCheckpointFileExist if a checkpoint file exists already
+// it returns err if running into any other exception
+func noneCheckpointFileExist(dir string, fileName string, totalSubtrieFiles int) error {
```

Comment:
Checkpointing will fail if some part file exists, that's why we need to ensure none Checkpoint part file exist.

Not only the header file, but also other part files must not exist.

Ideally, if some part file does exist, it's better that the error message contains the part file we found instead of just a boolean value. So that we could include that part file number into the error message for debugging purpose. 

In other words, errCheckpointFileExist will contain the part file number that we found. 

Since it's possible to hit an exception when checking file existence, I had to return two type of errors, and let caller to distinguish.

Reply:
It makes sense to log existing part files when none is expected.

I think it's cleaner to remove `noneCheckpointFileExist` since it is only called in `storeCheckpointV6`.  

Instead of wrapping `findCheckpointPartFiles` in `noneCheckpointFileExist()`, `storeCheckpointV6` can directly call `findCheckpointPartFiles` and return error if any part files are found (with enough information to log).

Thoughts?

Reply:
makes sense, updated